From 060b55ec3d56f6a836cadf484d017d1867bdc741 Mon Sep 17 00:00:00 2001
From: Dheeraj Kumar <quic_dhekum@quicinc.com>
Date: Wed, 31 May 2023 16:42:00 +0530
Subject: [PATCH] br-vlan-2
Signed-off-by: Dheeraj Kumar<quic_dhekum@quicinc.com>

---
 system-linux.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/system-linux.c b/system-linux.c
index 0f13a99..95d5560 100644
--- a/system-linux.c
+++ b/system-linux.c
@@ -1341,7 +1341,7 @@ int system_bridge_addbr(struct device *bridge, struct bridge_config *cfg)
 		nla_put_u64(msg, IFLA_BR_MCAST_STARTUP_QUERY_INTVL, val);
 	}
 
-	nla_put_u8(msg, IFLA_BR_VLAN_FILTERING, !!cfg->vlan_filtering);
+	//nla_put_u8(msg, IFLA_BR_VLAN_FILTERING, !!cfg->vlan_filtering);
 	nla_put_u16(msg, IFLA_BR_PRIORITY, cfg->priority);
 	nla_put_u32(msg, IFLA_BR_HELLO_TIME, sec_to_jiffies(cfg->hello_time));
 	nla_put_u32(msg, IFLA_BR_MAX_AGE, sec_to_jiffies(cfg->max_age));
-- 
2.17.1

