From d7fe0a14b673375115ee8e6fc3c65d2a4a5d5fe5 Mon Sep 17 00:00:00 2001
From: HuaWen CHEN <quic_huawenc@quicinc.com>
Date: Thu, 10 Oct 2024 20:18:32 +0800
Subject: [PATCH] netifd: skip rmnet_eth's parent device

Signed-off-by:quic_huawenc@quicinc.com
---
 device.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/device.c b/device.c
index 113614e..ff3907b 100644
--- a/device.c
+++ b/device.c
@@ -215,9 +215,11 @@ simple_device_set_state(struct device *dev, bool state)
 
 	pdev = dev->parent.dev;
 	if (state && !pdev) {
+	  if(strncmp(dev->ifname, "rmnet_eth", strlen("rmnet_eth")) != 0) {
 		pdev = system_if_get_parent(dev);
 		if (pdev)
 			device_add_user(&dev->parent, pdev);
+	  }
 	}
 
 	if (pdev) {
-- 
2.40.0.windows.1

