From 7ec6162c802ba4efe2ed1ac0a5631a360cb20d1a Mon Sep 17 00:00:00 2001
From: HuaWen CHEN <quic_huawenc@quicinc.com>
Date: Thu, 21 Nov 2024 13:45:39 +0800
Subject: [PATCH] netifd: skip rmnet's parent device

Signed-off-by: HuaWen CHEN <quic_huawenc@quicinc.com>
---
 device.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/device.c b/device.c
index 25d9dd8..4021754 100644
--- a/device.c
+++ b/device.c
@@ -197,9 +197,12 @@ simple_device_set_state(struct device *dev, bool state)
 
 	pdev = dev->parent.dev;
 	if (state && !pdev) {
+		if(strncmp(dev->ifname, "rmnet_eth", strlen("rmnet_eth")) != 0 &&
+		   strncmp(dev->ifname, "rmnet_data", strlen("rmnet_data")) != 0) {
 		pdev = system_if_get_parent(dev);
 		if (pdev)
 			device_add_user(&dev->parent, pdev);
+		}
 	}
 
 	if (pdev) {
-- 
2.40.0.windows.1

